src/object/srv_obj.c

```diff
@@ -1721,21 +1708,162 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc,
 
 	rc = obj_rw_complete(rpc, ioc, ioh, rc, dth);
 	if (iods_dup != NULL)
-		daos_iod_recx_free(iods_dup, orw->orw_nr);
+		daos_iod_recx_free(iods_dup, iods_nr);
 	return unlikely(rc == -DER_ALREADY) ? 0 : rc;
 }
 
+/* Extract local iods/offs/csums by orw_oid.id_shard from @orw */
 static int
-obj_local_rw(crt_rpc_t *rpc, struct obj_io_context *ioc,
-	     daos_iod_t *split_iods, struct dcs_iod_csums *split_csums,
-	     uint64_t *split_offs, struct dtx_handle *dth)
+obj_get_iods_offs_by_oid(daos_unit_oid_t uoid, struct obj_iod_array *iod_array,
+			 struct daos_oclass_attr *oca, uint64_t dkey_hash,
+			 uint32_t layout_ver, daos_iod_t **iods,
+			 uint64_t **offs, struct dcs_iod_csums **csums, uint32_t *nr)
 {
-	int	rc;
-	int	count = 0;
+	struct obj_shard_iod	*siod;
+	uint32_t		local_tgt;
+	uint32_t		oiod_nr;
+	int			i;
+	int 			idx = 0;
+	int			rc = 0;
+
+	oiod_nr = iod_array->oia_oiod_nr;
+	D_ASSERT(oiod_nr > 0);
+	if (oiod_nr > 1 || *iods == NULL) {
+		D_ALLOC_ARRAY(*iods, oiod_nr);
+		if (*iods == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		D_ALLOC_ARRAY(*offs, oiod_nr);
+		if (*offs == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
```

Comment:
some cleanup may need in the "out" label handling. (free mem, and zero the assigned pointer).

and is it possible to only allocate one piece of memory? in original code in obj_ec_rw_req_split(), it only allocate one time, now it may allocate multiple times in this critical IO path.
what's the main advantage to replace original obj_ec_rw_req_split() code? the code in this func basically do the same things as before right?

Reply:
oh, these will only be allocated if there multiple IODs for each RPC (if (oiod_nr > 1 || *iods == NULL)) , which is relatively rare case.  i.e. in most cases, it does not need allocate extra bytes here.

So pro:

1.  Removing split and re-packing RPC on the leader,  and also since we will need support multiple shards on the same target with OSA, changing the original ec_split to support this is a bit complicate as I understand, may need add extra dimension to support that.
2. Avoid memory allocation for most common case (1 iod per tgt), at least no allocation during DAOS handling.
3. Remove those layer violation handling in those RPC proc. 

Cons:
extra bytes will be needed to transfer between tgts, but it only add at most few daos_recx_t, so let's say, for EC16, 

old: transfer 16 bytes to each tgts, 
new: transfer 16 * 16  256 bytes to each tgts.

but the new method save the split and extra allocation on the leader. Hmm, probably the re-assembly can be smarter here to save some splitting/re-asembly on the leader, but we can discuss that later.

In my opinion this extra bytes does not add much latency, besides this extra latency will be buried over the whole forwarding/wait protocols anyway.